File: typescript/packages/core/src/multi-agent-router.ts
import type { AgentRuntime } from './runtime.js';
import type { AgentRunRequest, TerminalReason } from './types/domain.js';
import type { StreamEvent } from './types/contracts.js';

export interface AgentRouteRule {
  agentId: string;
  runtime: AgentRuntime;
  channel?: string;
  sessionPrefix?: string;
}

export interface RoutedAgentRequest extends AgentRunRequest {
  agentId?: string;
  channel?: string;
}

export class MultiAgentRouter {
  constructor(
    private readonly routes: AgentRouteRule[],
    private readonly fallback: AgentRuntime,
  ) {}

  resolve(request: RoutedAgentRequest): AgentRuntime {
    if (request.agentId) {
      const byId = this.routes.find((route) => route.agentId === request.agentId);
      if (byId) {
        return byId.runtime;
      }
    }

    for (const route of this.routes) {
      if (route.channel && request.channel && route.channel === request.channel) {
        return route.runtime;
      }
      if (route.sessionPrefix && request.sessionId.startsWith(route.sessionPrefix)) {
        return route.runtime;
      }
    }

    return this.fallback;
  }

  async *route(request: RoutedAgentRequest): AsyncGenerator<StreamEvent, TerminalReason, unknown> {
    const runtime = this.resolve(request);
    return yield* runtime.router.route(request);
  }
}

File: python/src/agent_framework/core/multi_agent_router.py
from __future__ import annotations

from dataclasses import dataclass

from agent_framework.runtime import AgentRuntime


@dataclass
class AgentRouteRule:
    agent_id: str
    runtime: AgentRuntime
    channel: str | None = None
    session_prefix: str | None = None


class MultiAgentRouter:
    def __init__(self, routes: list[AgentRouteRule], fallback: AgentRuntime) -> None:
        self._routes = routes
        self._fallback = fallback

    def resolve(self, *, session_id: str, agent_id: str | None = None, channel: str | None = None) -> AgentRuntime:
        if agent_id:
            for route in self._routes:
                if route.agent_id == agent_id:
                    return route.runtime
        for route in self._routes:
            if route.channel and channel and route.channel == channel:
                return route.runtime
            if route.session_prefix and session_id.startswith(route.session_prefix):
                return route.runtime
        return self._fallback
